src/librustc_mir/borrow_check.rs

```diff
@@ -615,29 +615,106 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
         let lvalue = self.base_path(lvalue_span.0);
 
         let maybe_uninits = &flow_state.uninits;
-        let move_data = maybe_uninits.base_results.operator().move_data();
-        if let Some(mpi) = self.move_path_for_lvalue(context, move_data, lvalue) {
+
+        // Bad scenarios:
+        //
+        // 1. Move of `a.b.c`, use of `a.b.c`
+        // 2. Move of `a.b.c`, use of `a.b.c.d` (without first reinitializing `a.b.c.d`)
+        // 3. Move of `a.b.c`, use of `a` or `a.b`
+        // 4. Uninitialized `(a.b.c: &_)`, use of `*a.b.c`; note that with
+        //    partial initialization support, one might have `a.x`
+        //    initialized but not `a.b`.
+        //
+        // OK scenarios:
+        //
+        // 5. Move of `a.b.c`, use of `a.b.d`
+        // 6. Uninitialized `a.x`, initialized `a.b`, use of `a.b`
+        // 7. Copied `(a.b: &_)`, use of `*(a.b).c`; note that `a.b`
+        //    must have been initialized for the use to be sound.
+        // 8. Move of `a.b.c` then reinit of `a.b.c.d`, use of `a.b.c.d`
+
+
+        // The dataflow tracks the shallow prefixes distinctly, in
+        // order to track substructure initialization separately from
+        // the whole structure.
+        //
+        // Therefore, it suffices to find the longest shallow prefix
+        // for which we have a MovePath (if any); the dataflow will
+        // take care of updating its state in response to
+        // super-structure moves and reinitializations.
+        //
+        // This is scenarios 1 and 2 above.
+
+        debug!("check_if_path_is_moved part1 lvalue: {:?}", lvalue);
+        if let Some(mpi) = self.move_path_closest_to(lvalue, PrefixSet::Shallow) {
             if maybe_uninits.curr_state.contains(&mpi) {
-                // find and report move(s) that could cause this to be uninitialized
                 self.report_use_of_moved(context, desired_action, lvalue_span);
-            } else {
-                // sanity check: initialized on *some* path, right?
-                assert!(flow_state.inits.curr_state.contains(&mpi));
+                return; // don't bother finding other problems.
+            }
+            // Only query longest prefix, not further ancestors;
+            // dataflow recurs on children when parents move (to
+            // support partial (re)inits).
+            //
+            // (I.e. querying parents breaks scenario 8; but may want
+            // to do such a query based on partial-init feature-gate.)
+        }
+
+        // A move of any shallow suffix of `lvalue` also interferes
+        // with an attempt to use `lvalue`. This is scenario 3 above.
+        //
+        // (Distinct from previous loop because `lvalue` does not
```

Comment:
nit: "previous loop" is confusing here. (The code has been internally refactored so that there is no longer a loop immediately present above, because its been factored into a helper method.)